SSA/Projects/Holor/Holor.lean

```diff
@@ -0,0 +1,101 @@
+import SSA.Core.WellTypedFramework
+import SSA.Core.Util
+import Mathlib.Data.Holor
+import Mathlib.Algebra.Algebra.Basic
+
+/- Holor is profunctorial, covariant in the values and contavariant in the indexes. -/
+
+/-- Mapping is covariant in the value -/
+def Holor.map (f : α → β) (h : Holor α ds) : Holor β ds :=
+  f ∘ h
+
+theorem Holor.map_functorial (f : α → β) (g : β → γ) (h : Holor α ds) :
+  Holor.map g (Holor.map f h) = Holor.map (g ∘ f) h :=
+  rfl
+
+/-- Reindexing is contravariant in the index -/
+def Holor.reindex (ix : HolorIndex ds₂ → HolorIndex ds₁) (h : Holor α ds₁) : Holor α ds₂ :=
+  λ ix₂ => h (ix ix₂)
+
+theorem Holor.reindex_functorial
+  (ix : HolorIndex ds₂ → HolorIndex ds₁) (iy : HolorIndex ds₃ → HolorIndex ds₂)
+  (h : Holor α ds₁) : Holor.reindex iy (Holor.reindex ix h) = Holor.reindex (ix ∘ iy) h := rfl
+
+
+/-- Pointiwse multiplication of holors. Lifts multiplicative structure into holor. -/
+def Holor.pointwise_mul [Mul α] (h₁ h₂ : Holor α ds) : Holor α ds := fun ix => (h₁ ix) * (h₂ ix)
+
+instance  [Mul α] : Mul (Holor α ds) where
+  mul := Holor.pointwise_mul
+
+theorem Holor.pointwise_mul_index [Mul α] (h₁ h₂ : Holor α ds) : 
+  (h₁ * h₂) i = h₁ i * h₂ i := rfl
+
+
+/-- Holor inherits algebraic structures --/
+instance [Semiring α] : Semiring (Holor α ds) := 
+  by delta Holor; infer_instance
+
+instance [CommSemiring α] : CommSemiring (Holor α ds) := 
+  by delta Holor; infer_instance
+
+instance [Ring α] : Ring (Holor α ds) := 
+  by delta Holor; infer_instance
+
+instance [CommRing α] : CommRing (Holor α ds) := 
+  by delta Holor; infer_instance
+
+instance [SMul M A] : SMul M (Holor A ds) := by 
+  delta Holor; infer_instance
+
+/-- Fill a holor with a constant value 'a'. -/
+def Holor.fill (a : α) : Holor α ds := fun _ => a
+
+-- @chrishughes24:
+-- given the map ralg : R -> Z(A), 
+-- I want to declare the map (r : R) -> (fill (ralg R) : Holor A ds).
+-- This will be the map that turns (Holor A ds) into an R-algebra.
+-- I'm not sure how to define this.
+instance [CommSemiring R] [Semiring A] [Algebra R A] : Algebra R (Holor A ds) :=  sorry
```

Comment:
@ChrisHughes24: 

given the map `ralg : R -> Z(A)`, I want to declare the map `(r : R) |-> (fill (ralg R) : Holor A ds)`.
This will be the map that turns `(Holor A ds)` into an `R`-algebra.
I'm not sure how to define this. In the spirit of "help a man to fish", could you tell me how I could figure this out for myself?